### PR TITLE
[Reland] Update mypy to 1.4.1

### PR DESCRIPTION
**Pull request by malfet**
This PR re-lands 
- [Typing] Fix PEP 484 Violation (#105022)
- Update mypy to 1.4.1 (#91983)

That were reverted due to the conflict with internal source repo.

Mostly fixes for PEP-484 violation (i.e. when default arg is set to None, but type is not annotated as optional)
Plus few real fixes:
  - Add missing `_get_upgraders_entry_map` to `torch/_C/__init__.pyi`
  - Add missing return statement to `torch._export. deserialize_graph`
  - Fix error message in `torch.ao.ns.fx.weight_utils.get_lstm_mod_weights`
  - Add assert it `torch/optim/optimizer.py` that Optional list is not None
TODO (in followup PR):
  - Fix erroneous `isinstance` check in `torch/ao/quantization/_pt2e/qat_utils.py`

Unrelated, to bypass CI failures due to the gcc9 dependency update in Ubuntu-18.04:
- Add hack to squash older libstdc++ from conda environment in favor one from OS to `.ci/docker/install_conda.sh`
- Update bazel cuda builds to focal, as with libstdc++-6.0.32 bazel builds loose the ability to catch exceptions (probably because they link with cupti statically, but I could not found where it is done)

### CODE REVIEW

**Review by atalman — APPROVED** [review]

**Review by albanD — APPROVED** [review]

File: torch/optim/optimizer.py [hunk 1]
```diff
@@ -418,14 +418,15 @@ def pack_group(group):
...
-                                                 param_groups: List[Dict[Any, Any]] = None, key=None) -> Tensor:
+    def _process_value_according_to_param_policy(param: Tensor, value: Tensor, param_id: Optional[int] = None,
+                                                 param_groups: Optional[List[Dict[Any, Any]]] = None, key=None) -> Tensor:
         # Floating-point types are a bit special here. They are the only ones
         # that are assumed to always match the type of params.
         # Make sure state['step'] is not casted https://github.com/pytorch/pytorch/issues/74424
         # UNLESS fused or capturable, see note [special device hosting for step]
         fused = False
         capturable = False
+        assert param_groups is not None
```
Comment (malfet) [inline torch/optim/optimizer.py:429]: @janeyx99 do you know what's the logic here? I.e. why `param_groups` is set to `None` by default, but then one tries to iterate over it on the next line (which would raise `TypeError` as None is not iterable)

**Review by Skylion007 — APPROVED** [review]
Woot! Finally

**Comment by malfet** [comment]
Docker attack :/

**Pull request closed without merging**